cql3/restrictions/single_table_query_restrictions.cc

```diff
@@ -7,3 +7,71 @@
  */
 
 #include "single_table_query_restrictions.hh"
+#include "cql3/expr/expression.hh"
+#include "cql3/restrictions/clustering_key_restrictions.hh"
+#include "seastar/core/on_internal_error.hh"
+
+namespace cql3 {
+namespace restrictions {
+using namespace expr;
+
+extern logging::logger rlogger;
+
+struct split_where_clause {
+    // Restrictions that only involve partition key columns
+    expression partition;
+
+    // Restrictions that only involve clustering key columns
+    expression clustering;
+
+    // All other restrictions from the where clause
+    expression other;
+};
+
+static split_where_clause get_split_where_clause(const expression& prepared_where_clause) {
+    std::vector<expression> partition;
+    std::vector<expression> clustering;
+    std::vector<expression> other;
+
+    for_each_boolean_factor(prepared_where_clause, [&](const expression& e) {
+        bool contains_partition = find_in_expression<column_value>(e, [](const column_value& cval) {
+                                      return cval.col->is_partition_key();
+                                  }) != nullptr;
+        bool contains_clustering = find_in_expression<column_value>(e, [](const column_value& cval) {
+                                       return cval.col->is_clustering_key();
+                                   }) != nullptr;
+        bool contains_other = find_in_expression<column_value>(e, [](const column_value& cval) {
+                                  return !cval.col->is_partition_key() && !cval.col->is_clustering_key();
+                              }) != nullptr;
+
+        if (contains_partition && !contains_clustering && !contains_other) {
+            partition.push_back(e);
+        } else if (!contains_partition && contains_clustering && !contains_other) {
+            clustering.push_back(e);
+        } else {
+            other.push_back(e);
+        }
+    });
```

Comment:
Looks like any random partition key restriction will end up in partition_key_restrictions. But I think it's better to move anything that doesn't fit to `other`. For example, `pk > 3` or `pl1 = ? AND pk3 = ?` (without pk2).

Reply:
But what if there aren't enough restrictions so that possible_lhs_values will work?

Anyway, I think this will resolve itself when we convert possible_lhs_values() to evalute(some_expression).